### PR TITLE
[SPARK-22686][SQL] DROP TABLE IF EXISTS should not show AnalysisException

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

During [SPARK-22488](https://github.com/apache/spark/pull/19713) to fix view resolution issue, there occurs a regression at `2.2.1` and `master` branch like the following. This PR fixes that.

```scala
scala> spark.version
res2: String = 2.2.1

scala> sql("DROP TABLE IF EXISTS t").show
17/12/04 21:01:06 WARN DropTableCommand: org.apache.spark.sql.AnalysisException:
Table or view not found: t;
org.apache.spark.sql.AnalysisException: Table or view not found: t;
```

## How was this patch tested?

Manual.

### CODE REVIEW

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 1]
```diff
@@ -206,7 +206,8 @@ case class DropTableCommand(
     try {
       sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName))
     } catch {
-      case _: NoSuchTableException if ifExists =>
+      case ae: AnalysisException
+        if ifExists && ae.cause.nonEmpty && ae.getCause.isInstanceOf[NoSuchTableException] =>
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: can we follow https://github.com/apache/spark/pull/19713/files#diff-bc55b5f76add105ec32ae4107075b278R57

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]:
```
scala> spark.version
res2: String = 2.3.0-SNAPSHOT

scala> sql("DROP TABLE IF EXISTS t")
res3: org.apache.spark.sql.DataFrame = []

scala> sql("DROP TABLE IF EXISTS t")
res4: org.apache.spark.sql.DataFrame = []
```

Unable to reproduce it.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: it's in the log, which means it's not a serious bug.

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I can see it in 2.2.1. It is a warning message.

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: hmmm, it's fixed in master?

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: @dongjoon-hyun can you verify it? thanks!

Comment (gatorsmile) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I can reproduce it in master now.

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: Yep. It exists in master, too.
I'll update like the example by @cloud-fan .

**Comment by dongjoon-hyun** [comment]
Retest this please.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 2]
```diff
@@ -203,14 +203,16 @@ case class DropTableCommand(
         case _ =>
       }
     }
-    try {
-      sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName))
-    } catch {
-      case _: NoSuchTableException if ifExists =>
-      case NonFatal(e) => log.warn(e.toString, e)
+
+    if (!ifExists || catalog.tableExists(tableName)) {
```
Comment (cloud-fan) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I feel it's more readable to write
```
if (catalog.tableExists(tableName)) {
  ... do the drop
} else if (ifExists) {
  // do nothing
} else {
  // log.warn("The table to drop does not exist: " +tableName)
}
```

Comment (cloud-fan) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: We can also update the UNCACHE TABLE, cc @gatorsmile

Comment (dongjoon-hyun) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: I'll update soon.

**Comment by dongjoon-hyun** [comment]
I added a VIEW case and we need to throw `AnalysisException` if `ifExists=false`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala [hunk 3]
```diff
@@ -203,14 +203,20 @@ case class DropTableCommand(
...
+    if (catalog.isTemporaryTable(tableName) || catalog.tableExists(tableName)) {
+      try {
+        sparkSession.sharedState.cacheManager.uncacheQuery(sparkSession.table(tableName))
+      } catch {
+        case NonFatal(e) => log.warn(e.toString, e)
+      }
+      catalog.refreshTable(tableName)
+      catalog.dropTable(tableName, ifExists, purge)
+    } else if (ifExists) {
+      // no-op
```
Comment (gatorsmile) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/command/ddl.scala]: ?

**Comment by gatorsmile** [comment]
LGTM pending Jenkins

**Comment by dongjoon-hyun** [comment]
Thank you, @gatorsmile .

**Comment by cloud-fan** [comment]
thanks, merging to master/2.2!

**Comment by dongjoon-hyun** [comment]
Thank you, @cloud-fan and @gatorsmile !

**Pull request closed without merging**